source/Ship.cpp

```diff
@@ -4057,3 +4354,319 @@ void Ship::CreateSparks(vector<Visual> &visuals, const Effect *effect, double am
+}
+
+
+
+// A small check to see if a ship may request help.  Primarily for player escort
+// fleet but can be used by any ship.
+bool Ship::MayRequestHelp() const
+{
+	return (!isTankerCarrier && IsFuelLow()) || IsEnergyLow() || IsDisabled();
+}
+
+
+
+// Perform calculations across all escorts to set private variables
+// escortsHaveOneJump, isEnemyInEscortSystem, isEscortsFullOfFuel, and
+// refuelMissionNpcEscort.  Warning this is an expensive calculation if done too
+// often.
+void Ship::UpdateEscortsState(shared_ptr<Ship> flagship, const vector<weak_ptr<Ship>> allEscorts)
+{
+	bool updatedEscortsFullOfFuel = false;
+	bool updatedEnemyInEscortSystem = false;
+	bool updateEscortsHaveOneJump = false;
+	bool updateRefuelMissionNpcEscort = false;
+
+	const Government *gov = flagship->GetGovernment();
+	// Update flagship state based on escorts
+	for(const weak_ptr<Ship> &ptr : allEscorts)
+	{
+		shared_ptr<Ship> escort = ptr.lock();
+		if(!IsEscortedBy(flagship, escort))
+			continue;
+
+		if(!escort->CanBeCarried() && escort->GetEscorts().size())
+			UpdateEscortsState(flagship, escort->GetEscorts());
+
+		if(escort->CanBeCarried() && !escort->GetShipToAssist())
+			flagship->maxCarriedShipFuel = max(flagship->maxCarriedShipFuel, escort->Fuel());
+
+		if(!updatedEscortsFullOfFuel && escort->IsFuelLow() && !escort->CanBeCarried() && !escort->IsTankerCarrier())
+		{
+			flagship->isEscortsFullOfFuel = false;
+			updatedEscortsFullOfFuel = true;
+		}
+
+		if(!updatedEnemyInEscortSystem)
+		{
+			shared_ptr<const Ship> escortTarget = escort->GetTargetShip();
+			if(escortTarget)
+				if(gov->IsEnemy(escortTarget->GetGovernment()))
+				{
+					flagship->isEnemyInEscortSystem = true;
+					updatedEnemyInEscortSystem = true;
+				}
+		}
+
+		if(!updateEscortsHaveOneJump)
+			if(escort->JumpFuelMissing())
+			{
+				flagship->escortsHaveOneJump = false;
+				updateEscortsHaveOneJump = true;
+			}
+
+		if(!updateRefuelMissionNpcEscort)
+			if(escort->IsYours() && escort->IsFuelLow() && !escort->CanBeCarried())
+			{
+				flagship->refuelMissionNpcEscort = false;
+				updateRefuelMissionNpcEscort = true;
+			}
+	}
+	if(!updatedEscortsFullOfFuel)
+		flagship->isEscortsFullOfFuel = true;
+	if(!updatedEnemyInEscortSystem)
+		flagship->isEnemyInEscortSystem = false;
+	if(!updateEscortsHaveOneJump)
+		flagship->escortsHaveOneJump = true;
+	if(!updateRefuelMissionNpcEscort)
+		flagship->refuelMissionNpcEscort = true;
+}
+
+
+
+// Update the carried ship state based on this ship state.
+void Ship::UpdateEscortsState(shared_ptr<Ship> other)
+{
+	if(other != shared_from_this())
+	{
+		other->isEscortsFullOfFuel = isEscortsFullOfFuel;
+		other->isEnemyInEscortSystem =	isEnemyInEscortSystem;
+		other->escortsHaveOneJump = escortsHaveOneJump;
+		other->refuelMissionNpcEscort = refuelMissionNpcEscort;
+	}
+
+	// Set other minimum hull to the calculated value.
+	other->minimumHull = other->CalculateMinimumHull();
+	other->isDisabled = CalculateIsDisabled();
```

Comment:
This is a mistake.  other should be calculated from other and not this ship.  This causes IsDisabled to stutter when playing.